### PR TITLE
Changing cirrina.websockets to a dict

### PR DESCRIPTION
**Pull request by arcenik**
Changing cirrina.websockets to a dict to allow sending non-broadcast messages.

### CODE REVIEW

File: cirrina/server.py [hunk 1]
```diff
@@ -380,9 +380,8 @@ def websocket_broadcast(self, msg):
         """
         Broadcast a message to all websocket connections.
         """
-        for websocket in self.websockets:
-            # FIXME: use array
-            websocket.send_str('{"status": 200, "message": %s}'%json.dumps(msg))
+        for key in self.websockets:
```
Comment (timofurrer) [inline cirrina/server.py]: Use `dict.values()`to iterate over the values of the dict.

File: cirrina/server.py [hunk 2]
```diff
@@ -380,9 +380,8 @@ def websocket_broadcast(self, msg):
...
+            self.websockets[key].send_str('{"status": 200, "message": %s}'%json.dumps(msg))
```
Comment (timofurrer) [inline cirrina/server.py]: Use `format()` instead of `%`-formatting

File: cirrina/server.py [hunk 3]
```diff
@@ -425,7 +424,7 @@ def _ws_handler(self, request):
             websocket.close()
             return websocket
 
-        self.websockets.append(websocket)
+        self.websockets[websocket.__hash__()] = websocket
```
Comment (timofurrer) [inline cirrina/server.py:427]: If an object is [hashable](https://docs.python.org/3/glossary.html#term-hashable) you can use it directly has the key for the dict. No need to call `__hash__`explicitly.

File: cirrina/server.py [hunk 4]
```diff
@@ -444,7 +443,7 @@ def _ws_handler(self, request):
 
             yield from asyncio.sleep(0.1)
 
-        self.websockets.remove(websocket)
+        self.websockets.pop(websocket.__hash__())
```
Comment (timofurrer) [inline cirrina/server.py:446]: same

**Review by timofurrer — CHANGES_REQUESTED** [review]

File: cirrina/server.py [hunk 5]
```diff
@@ -127,8 +127,8 @@ def _stop(self):
         logger.debug('Stopping cirrina server...')
         for handler in self.shutdown_handlers:
             handler()
-        for ws in self.websockets:
-            ws.close()
+        for key in self.websockets:
```
Comment (timofurrer) [inline cirrina/server.py]: Use `dict.values()` as well here.

**Comment by timofurrer** [comment]
Why do you want to do this? I do not see a benefit of using a `dict` here, because we never search for a specific websocket object. The dict key in this implementation would be the hash of the websocket object - and we only know this hash if we already have a websocket object. So, I don't quiet see benefit yet.

> ... to allow sending non-broadcast messages.

I've just read this comment. 
However, the last point above is still a question to me. How do you want to lookup a websocket object in this dict if you need the object's hash but don't have a websocket object yet?

**Comment by neolynx** [comment]
I like the idea, however we need to find what the key is supposed to be. I was thinking of using the socket file descriptor as it is unique. But then, it's still equal to an array. we cannot use the username as key, as a user might be logged in in multiple browsers ans is not unique. 
I left it as array because of lack of ideas :)

maybe the websocket mesages should enable some simple pub sub pattern, which could be used to create "channels" to broadcast to. Other frameworks (redis?) might be doing that already...

File: cirrina/server.py (hunk 3 above)
Comment (arcenik) [reply cirrina/server.py:427]: No I need a string so I can use it with a key/value store like Redis

**Comment by arcenik** [comment]
Here is the usage scenario :
* user#1 list display a project
* user#1 websocket is registered in a diffusion group (__hash__ key in redis)
* user#2 change the project
* by using the diffusion group, user#1 is notified and his display is refreshed

File: cirrina/server.py (hunk 1 above)
Comment (arcenik) [reply cirrina/server.py]: done

**Pull request closed without merging**